File: supervised_learning/DT.R
# DECISION TREES IN R
library(tidyverse)
library(rpart)
library(rattle)
library(caret)
data <- read_csv("supervised_learning/data/firms_3883.csv")

#Decision tree for MIXED data, without normalization
#Outcome: Dividend payment (Y/N) by dividend yield

set.seed(9999)

data_dt <- data%>%
  select(Sector, Exchange, MarketCapitalization, BookValue, EPS, RevenuePerShareTTM,
         ProfitMargin, OperatingMarginTTM, ReturnOnAssetsTTM, ReturnOnEquityTTM, 
         RevenueTTM, QuarterlyEarningsGrowthYOY, QuarterlyRevenueGrowthYOY, DividendYield)%>%
  mutate(Dividend = DividendYield>0)%>%
  select(-DividendYield)

train <- data_dt%>%sample_frac(0.75, replace = FALSE)
test <- data_dt%>%anti_join(train)


data_dt


#Data appears mostly balanced
summary(train)
summary(test)

#Store label and remove from test set
test_label <- test$Dividend
test <- test%>%select(-Dividend)


#Train the Decision Tree
dt1 <- rpart(Dividend ~ ., data = train, method = "class", cp =0.003)
summary(dt1)

#Visualize the tree
fancyRpartPlot(dt1)

#Predict Labels
dt1_pred = predict(dt1, test, type = "class")
test_label <- as.factor(test_label)

#Confusion matrix
confusionMatrix(dt1_pred, test_label)




#Decision tree for discretized data
#Import discretized data from ARM section?
#Outcome variable: DIVIDEND


#Discretized: 
# Select attributes
data_dt2 <- data%>%
  select(Sector, Sector_NDAQ, Exchange, MarketCapitalization, BookValue, EPS, RevenuePerShareTTM,
         ProfitMargin, OperatingMarginTTM, ReturnOnAssetsTTM, ReturnOnEquityTTM, 
         QuarterlyEarningsGrowthYOY, QuarterlyRevenueGrowthYOY, DividendYield)



#Discretize variables:

#Dividend, binary (True if div paid, else false)
data_dt2 <- data_dt2%>%
  mutate(Dividend = DividendYield>0)%>%
  select(-DividendYield)

#EPS, pos or negative
# Negative - Zero - Positive : Profit Margin, Operating Margin, Earnings Growth, Revenue Growth, ROA, ROE

data_dt2 <- data_dt2%>%
  mutate(EPS = EPS>0)%>% 
  mutate(ProfitMargin = if_else(ProfitMargin < 0, 
                                "Negative", 
                                if_else(ProfitMargin == 0, 
                                        "Zero",
                                        "Positive")),
         OperatingMarginTTM = if_else(OperatingMarginTTM < 0, 
                                      "Negative", 
                                      if_else(OperatingMarginTTM == 0, 
                                              "Zero",
                                              "Positive")),
         QuarterlyEarningsGrowthYOY = if_else(QuarterlyEarningsGrowthYOY < 0, 
                                              "Negative", 
                                              if_else(QuarterlyEarningsGrowthYOY == 0, 
                                                      "Zero",
                                                      "Positive")),
         QuarterlyRevenueGrowthYOY = if_else(QuarterlyRevenueGrowthYOY < 0, 
                                              "Negative", 
                                              if_else(QuarterlyRevenueGrowthYOY == 0, 
                                                      "Zero",
                                                      "Positive")),
         ReturnOnAssetsTTM = if_else(ReturnOnAssetsTTM < 0, 
                                    "Negative", 
                                    if_else(ReturnOnAssetsTTM == 0, 
                                            "Zero",
                                            "Positive")),
         ReturnOnEquityTTM = if_else(ReturnOnEquityTTM < 0, 
                                     "Negative", 
                                     if_else(ReturnOnEquityTTM == 0, 
                                             "Zero",
                                             "Positive")))


#Split dataset
train2 <- data_dt2%>%sample_frac(0.75, replace = FALSE)
test2 <- data_dt2%>%anti_join(train2)


test2_label <- test2$Dividend
test2 <- test2%>%select(-Dividend)


#Run 2nd DT in R with discrete data

dt2 <- rpart(Dividend ~ ., data = train2, method = "class", cp =0.003)
summary(dt1)

fancyRpartPlot(dt2)

dt2_pred = predict(dt2, test2, type = "class")
test2_label <- as.factor(test2_label)

confusionMatrix(dt2_pred, test2_label)

File: supervised_learning/DT.py
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.metrics import confusion_matrix, classification_report
import seaborn as sns

data = pd.read_csv("data/firms_3883.csv")

# Numeric Data DT - Predicting positive or negative EPS

X = data[['MarketCapitalization', 'BookValue','DividendYield', 'RevenuePerShareTTM',
          'ProfitMargin', 'OperatingMarginTTM', 'ReturnOnAssetsTTM', 'ReturnOnEquityTTM',
          'QuarterlyEarningsGrowthYOY','QuarterlyRevenueGrowthYOY']]

y = data['EPS']>=0

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 40)

dt = tree.DecisionTreeClassifier()
dt.fit(X_train, y_train)

y_predicted = dt.predict(X_test)

confmatrix = confusion_matrix(y_test, y_predicted)
columns = ['Predicted Positive', 'Predicted Positive']
rows = ['Actual Positive', 'Actual Negative']
fig, ax = plt.subplots(figsize = (8,6))
sns.heatmap(confmatrix, annot = True, fmt = 'd', xticklabels = columns, yticklabels = rows, ax = ax)
plt.title("Confusion Matrix")
plt.savefig("confmatrix1.png")



print(classification_report(y_test, y_predicted))

fig = plt.figure(figsize = (100,80))
e = tree.plot_tree(dt,
                   fontsize = 8,
                   feature_names=X.columns,
                   class_names=True,
                   filled = True)
plt.savefig("dt1.png")


